### ISSUE TITLE
Support controlling listener calling order with `ROBOT_LISTENER_PRIORITY` attribute

### ISSUE THREAD

**Issue by ElColmo** [robotframework/robotframework#3473]
I'm on RobotFramework 3.1.2 on Linux.

I have a Library (MyLibrary) with a built-in listener, implemented via {code}self.ROBOT_LIBRARY_LISTENER=self{code}.

I also have an external listener, implemented via poc-listener.py.

I want to be able to specify the Library's built-in listener to be notified about the start and end events *before* the  external listener is notified. What happens in practice is that the external listener is always notified first. I use the following command line:
{code}robot --listener poc-listener.py testcases{code}

Note that "testcases" has a test suite file which specifies the library import:
    *** Settings ***
    Library MyLibrary

This same behaviour happens even if I disable the built-in listener, and specify both listeners via the command line, using:
    robot --listener poc-listener.py --listener MyLibrary testcases

The only way I've been able to find to control the order in which the listeners execute is by specifying *both* listeners in the Library, using 
    self.ROBOT_LIBRARY_LISTENER={self, "poc-listener.py"}
and by calling robot without any listeners on the command line: 
    robot testcases.

Obviously this isn't ideal, since 
- I'd be forced to add parameters to the library, to specify the external listeners
- this completely bypasses Robot's built-in method of specifying the listeners (via the command line)
- it breaks encapulation, since, in general most Libraries have no business knowing about what other listeners have been specified

What I'd suggest is a mechanism to allow the order of listeners to be specified, where one of the listeners involved is a listener specified within a Library itself, using
    self.ROBOT_LIBRARY_LISTENER=self

**Comment by pekkaklarck** [comment]
It's true that listeners registered from the command line with `--listener` are notified about events before listeners registered by libraries. It ought to be possible to control the relative order of these listeners, though. If you use, for example, `--listener A --listener B`, then `A` ought to be notified before `B`. Same way, library listeners ought to be notified in the order libraries are imported.

Someone else has mentioned this same problem (was it you @Snooz82?) and it would be nice to have a solution. There are, however, some problems related to this:

- I don't have any good design idea how listeners could control this.
- This functionality is only needed if you have both normal listeners and library listeners. 
- Implementation would be pretty big task due to how listeners are implemented currently.

Due to the above problems it's very unlikely this feature will be implemented in the near future. We could, however, consider changing the functionality so that library listeners are notified first and "normal" listeners only after them. That ought to be a trivial change but due to the change being backwards incompatible we cannot do it lightly. For me library listeners being notified first sound good. Any other opinions?

**Comment by Snooz82** [comment]
I think it would be bad to just change the order!

one would like the cli be first, others would like the libraries be first.


My idea was that a Listener could define a "Priority" .

like `ROBOT_LISTENER_PRIORITY = 0` 
Or `ROBOT_LISTENER_PRIORITY = 100` 
Or a negative value.

If not defined, priority would be 0 and therefore listeners with would be called in initialization order.
That would mean, cli-listener first in order then lib-listener in their import order.
This is in general straight forward.

Same behavior we have with variables: first come first serve. And cli is before Resource or variable file imports. Therefore CLI shall have higher priority.


BUT:
If someone would want a specific listener to be called very first, you could give it a higher priority like 9999 . 
DataDriver or other modification listeners are such a group of things that want to be called early.

on the other hand, listeners that report a Status of a keyword or TestCase to another system, want to be called last. So you define a negative priority.

similar to Z-Index on web pages.

**Comment by Snooz82** [comment]
Example of reasons for an order:

one Listener removes confidential Data from the Messages.
Another listener sends all Keywords to a Reporting system.

The Cleaner should run first!

other thing:
A CLI-Listener sends all Tests to a Dashboard, when Suite starts and reports all status to these test cases once each is done.
A Library Listener like DataDriver or so, creates new test cases.
Then you have the problem, that there are test cases execute, which the library listener has never seen in suite_start

**Comment by Snooz82** [comment]
@jleigh100 
Could you maybe explain your use case a little bit more concrete, so that other would understand the need a bit better?

**Comment by pekkaklarck** [comment]
I don't have a strong opinion should listener registered from the CLI or library listeners be registered first. As I commented earlier, the order could be changed, but due to backwards compatibility reasons that change cannot be done lightly. It could only be done in a major version like RF 7.0 and even then there should be concrete benefits. As @Snooz82 pointed out above, there are use cases for both approaches, so if we change the order someone will affected negatively.

I guess the best solution is making it possible for listeners to specify the order like @Snooz82 proposed above. I like his design with `ROBOT_LISTENER_PRIORITY` that can have positive or negative values and zero being the default as well. That makes it easy to state that a certain listener should be used before or after those with the default priority. It's also possible to fine-tune the order of listeners but that requires co-operation between listeners.

Setting priorities ought to be relatively simple so that support could possibly be implemented already in RF 6.1. I assign this issue to that milestone and we can then prioritize it along with other issues when we actually start RF 6.1 development.

**Comment by jleigh100** [comment]
That sounds good - thank you!

As for context, we have a web app which allows users to run Robot Framework scripts, and we always use a 'normal' listener (always called with --listener) which includes an `end_test` which makes a call to the web app, so a progress bar can be displayed there. Some of our Robot Framework scripts contain tests which test hardware, which, at the end of a test, can take a few seconds to actually stop. Only in those scripts, we include a listener as a library, which also has an `end_test` which waits for the hardware to have stopped before telling the web app the test has finished. However, the 'normal' listener's `end_test` is called first, which means the web app shows the test is completed before the hardware is actually free. Due to logic in the front-end, this has implications such as the next Robot Framework script starting before the hardware is actually free.

**Comment by d-biehl** [comment]
comment from roadmap talk:

Maybe we should not say priority, I think it would be better to precedence.

"precedence" is more formal and less common than "priority". There is a small difference in meaning: priority refers to importance, whereas precedence refers to sequence.

**Comment by pekkaklarck** [comment]
Good point about terminology above. Perhaps "order" would be even better than "precedence" or "priority"? The plan is to use index similarly as when defining [z-order](https://developer.mozilla.org/en-US/docs/Web/CSS/z-index) in CSS and using same terminology as there feels like a good idea.

If we go with "order", we still need to decide is the special attribute controlling the order named `ROBOT_LISTENER_ORDER` or `ROBOT_LISTENER_INDEX`. The latter would match the `z-index` CSS property, but the former is at least somewhat self explanatory and probably better.

**Comment by pekkaklarck** [comment]
This is now done. I decided to use `ROBOT_LISTENER_ORDER` as the name of the special attribute controlling the order.

**Comment by pekkaklarck** [comment]
The attribute name was change to `ROBOT_LISTENER_PRIORITY` in the end. The decision was based on the discussion on our Slack and the main motivation was that although the attribute is used for specifying the order, a single listener only sets its own priority, not the whole order.
